app/services/html/parser.rb

```diff
@@ -0,0 +1,284 @@
+      end
+
+      @html = doc.to_html
+
+      self
+    end
+
+    def add_control_class_to_codeblock
+      doc = Nokogiri::HTML.fragment(@html)
+
+      doc.search("div.highlight").each do |codeblock|
+        codeblock.add_class("js-code-highlight")
+      end
+
+      @html = doc.to_html
+
+      self
+    end
+
+    def add_control_panel_to_codeblock
+      doc = Nokogiri::HTML.fragment(@html)
+
+      doc.search("div.highlight").each do |codeblock|
+        codeblock.add_child('<div class="highlight__panel js-actions-panel"></div>')
+      end
+
+      @html = doc.to_html
+
+      self
+    end
+
+    def add_fullscreen_button_to_panel
+      on_title = "Enter fullscreen mode"
+      on_cls = "highlight-action highlight-action--fullscreen-on"
+      icon_fullscreen_on = inline_svg_tag("fullscreen-on.svg", class: on_cls, title: on_title)
+      off_title = "Exit fullscreen mode"
+      off_cls = "highlight-action highlight-action--fullscreen-off"
+      icon_fullscreen_off = inline_svg_tag("fullscreen-off.svg", class: off_cls, title: off_title)
+      doc = Nokogiri::HTML.fragment(@html)
+      doc.search("div.highlight__panel").each do |codeblock|
+        fullscreen_action = <<~HTML
+          <div class="highlight__panel-action js-fullscreen-code-action">
+              #{icon_fullscreen_on}
+              #{icon_fullscreen_off}
+          </div>
+        HTML
+
+        codeblock.add_child(fullscreen_action)
+      end
+
+      @html = doc.to_html
+
+      self
+    end
+
+    def wrap_all_tables
+      doc = Nokogiri::HTML.fragment(@html)
+      doc.search("table").each { |table| table.swap("<div class='table-wrapper-paragraph'>#{table}</div>") }
+      @html = doc.to_html
+
+      self
+    end
+
+    def remove_empty_paragraphs
+      doc = Nokogiri::HTML.fragment(@html)
+      doc.css("p").select { |paragraph| all_children_are_blank?(paragraph) }.each(&:remove)
+      @html = doc.to_html
+
+      self
+    end
+
+    def escape_colon_emojis_in_codeblock
+      html_doc = Nokogiri::HTML.fragment(@html)
+
+      html_doc.children.each do |el|
+        next if el.name == "code"
+
+        if el.search("code").empty?
+          if el.parent.present?
+            parsed_html = Html::Parser.new(el.to_html).parse_emojis.html
+            el.swap(parsed_html)
+          end
+        else
+          el.children = self.class.new(el.children.to_html)
+            .escape_colon_emojis_in_codeblock
+            .html
+        end
+      end
+
+      @html = html_doc.to_html
+
+      self
+    end
+
+    def unescape_raw_tag_in_codeblocks
+      return self if @html.blank?
+
+      @html.gsub!(RAW_TAG, "{% raw %}")
+      @html.gsub!(END_RAW_TAG, "{% endraw %}")
+      html_doc = Nokogiri::HTML(@html)
+      html_doc.xpath("//body/div/pre/code").each do |codeblock|
+        next unless codeblock.content.include?(RAW_TAG) || codeblock.content.include?(END_RAW_TAG)
```

Comment:
Doing this wasn't my first choice 😅 

When I changed the name of this file to `Html::Parser`, it caused this line to be indented further because we added `module Html` at the top. That caused a rubocop violation for the line then being too long.

Then, I moved the logic to a method, something like:
```ruby
next unless codeblock_includes_raw_tags?(codeblock.content)

def codeblock_includes_raw_tags?(content)
  content.include?("{----% raw %----}") || content.include?("{----% endraw %----}")
end
```

Andddddd then codeclimate complained because the file contained 21 methods which is over the 20 method limit. I try to avoid "hacks" (i.e. disabling Rubocop and/or ignoring codeclimate) so this where I landed to satisfy both 🙈 .